### PR TITLE
feat: emergency stop

### PR DESCRIPTION
**Pull request by graykode**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,7 +18,7 @@
     "typescript": "^5.0.0"
   },
   "dependencies": {
-    "@clober/v2-sdk": "^0.0.57",
+    "@clober/v2-sdk": "^0.0.58",
```
Comment (graykode) [inline package.json:21]: TODO update to 0.0.59

File: model/clober-market-maker.ts [hunk 2]
```diff
@@ -190,6 +193,30 @@ export class CloberMarketMaker {
...
+      if (slackClient) {
+        await slackClient.error({
+          message: 'Error in market making',
+          error: `Off-chain block number ${offChainBlockNumber} is too behind of on-chain block number ${onChainBlockNumber}`,
+        })
+      }
+      this.isEmergencyStop = true
+      await this.sleep(30 * 1000)
+      throw new Error(
+        `Off-chain block number ${offChainBlockNumber} is too ahead of on-chain block number ${onChainBlockNumber}`,
```
Comment (Dorvin) [inline model/clober-market-maker.ts:215]: `Off-chain block number ${offChainBlockNumber} is too behind of on-chain block number ${onChainBlockNumber}`

Comment (Dorvin) [inline model/clober-market-maker.ts:215]:
```suggestion
        `Off-chain block number ${offChainBlockNumber} is too behind of on-chain block number ${onChainBlockNumber}`,
```

File: model/clober-market-maker.ts [hunk 3]
```diff
@@ -485,6 +513,8 @@ export class CloberMarketMaker {
         market,
         ...newEpoch,
       })
+      // always make when new epoch
```
Comment (Dorvin) [inline model/clober-market-maker.ts:516]: how about first epoch? need to initialize...!

File: model/clober-market-maker.ts [hunk 4]
```diff
@@ -190,6 +193,30 @@ export class CloberMarketMaker {
...
+          message: 'Error in market making',
+          error: `Off-chain block number ${offChainBlockNumber} is too behind of on-chain block number ${onChainBlockNumber}`,
+        })
+      }
+      this.isEmergencyStop = true
+      await this.sleep(30 * 1000)
+      throw new Error(
+        `Off-chain block number ${offChainBlockNumber} is too ahead of on-chain block number ${onChainBlockNumber}`,
+      )
+    }
```
Comment (Dorvin) [inline model/clober-market-maker.ts:217]:
```suggestion
    } else {
        this.isEmergencyStop = false
    }
```

File: model/clober-market-maker.ts [hunk 5]
```diff
@@ -190,6 +193,30 @@ export class CloberMarketMaker {
     this.initialized = true
   }
 
+  async emergencyStopCheck(makeBlockInterval: number) {
```
Comment (Dorvin) [inline model/clober-market-maker.ts:196]:
```suggestion
  async emergencyStopCheck(blockDelayThreshold: number) {
```

File: model/clober-market-maker.ts [hunk 6]
```diff
@@ -199,6 +226,7 @@ export class CloberMarketMaker {
     while (true) {
       try {
         await Promise.all([
+          this.emergencyStopCheck(this.config.makeBlockInterval),
```
Comment (Dorvin) [inline model/clober-market-maker.ts:229]: need to 'continue' when the emergency stop is triggered, and check again as soon as possible

File: model/clober-market-maker.ts (hunk 3 above)
Comment (graykode) [reply model/clober-market-maker.ts:516]: https://github.com/clober-dex/clober-market-maker/pull/59/commits/f421d6f2203dae538526e1ec3250e4e472c1242f 

@Dorvin

**Review by Dorvin — APPROVED** [review]

**Pull request merged**